src/EditorFeatures/CSharpTest/CodeRefactorings/UseRecursivePatterns/UseRecursivePatternsRefactoringTests.cs

```diff
@@ -0,0 +1,146 @@
+        [InlineData("this.P1 == 1 && this.P2 == 2", "this is { P1: 1, P2: 2 }")]
+        [InlineData("this.P1 < 1 && this.P2 <= 2", "this is { P1: < 1, P2: <= 2 }")]
+        [InlineData("this.P1 > 1 && this.P2 >= 2", "this is { P1: > 1, P2: >= 2 }")]
+        [InlineData("this.P1 != 1 && this.P2 != 2", "this is { P1: not 1, P2: not 2 }")]
+        [InlineData("this.P1 == 1 && this.CP2.P3 == 3", "this is { P1: 1, CP2: { P3: 3 } }")]
+        [InlineData("this.CP1.P1 == 1 && this.CP1.CP2.P3 == 3", "this.CP1 is { P1: 1, CP2: { P3: 3 } }")]
+        public async Task TestLogicalAndExpression(string actual, string expected)
+        {
+            await VerifyAsync(WrapInIfStatement(actual, "&&"), WrapInIfStatement(expected, "&&"));
+        }
+
+        [Theory]
+        [InlineData("this.CP1 is var c [||]&& c.P1 == 0", "this.CP1 is { P1: 0 } c")]
+        [InlineData("this.P1 == 1 && this.CP1 is var c [||]&& c.P1 == 0 && this.P2 == 2", "this.P1 == 1 && this.CP1 is { P1: 0 } c && this.P2 == 2")]
+        [InlineData("this.P1 == 1 && this.CP1.P1 == 1 [||]&& this.CP1.CP2.P3 == 3", "this.P1 == 1 && this.CP1 is { P1: 1, CP2: { P3: 3 } }")]
+        [InlineData("this.P1 == 1 && this.CP1.P1 == 1 [||]&& this.CP1.CP2.P3 == 3 && this.P2 == 2", "this.P1 == 1 && this.CP1 is { P1: 1, CP2: { P3: 3 } } && this.P2 == 2")]
+        public async Task TestLogicalAndExpressionChildren(string actual, string expected)
+        {
+            await VerifyAsync(WrapInIfStatement(actual, entry: null), WrapInIfStatement(expected, entry: null));
+        }
+
+        [Theory]
+        [InlineData("NS.C.SCP1 == null && NS.C.SCP2 == null")]
+        [InlineData("NS.C.SCP1.P1 == 1 && NS.C.SCP2.P1 == 2")]
+        public async Task TestLogicalAndExpressionMissing(string actual)
+        {
+            await VerifyMissingAsync(WrapInIfStatement(actual, "&&"));
+        }
+
+        [Theory]
+        [InlineData("{ CP1: var c } when c.P1 is 1", "{ CP1: { P1: 1 } c }")]
+        [InlineData("{ CP1: var c } when c.P1 == 1", "{ CP1: { P1: 1 } c }")]
+        [InlineData("{ CP1: var c } when c.P1 == 1 && c.P2 == 2", "{ CP1: { P1: 1 } c } when c.P2 == 2")]
+        [InlineData("{ CP1: C c } when c.P1 == 1", "{ CP1: C { P1: 1 } c }")]
+        [InlineData("{ CP1: C { P2: 2 } c } when c.P1 == 1", "{ CP1: C { P2: 2, P1: 1 } c }")]
+        public async Task TestWhenClause(string actual, string expected)
+        {
+            await VerifyAsync(WrapInSwitchArm(actual, "when"), WrapInSwitchArm(expected, "when"));
+            await VerifyAsync(WrapInSwitchArm(actual, "=>"), WrapInSwitchArm(expected, "=>"));
+            await VerifyAsync(WrapInSwitchLabel(actual, "when"), WrapInSwitchLabel(expected, "when"));
+            await VerifyAsync(WrapInSwitchLabel(actual, "case"), WrapInSwitchLabel(expected, "case"));
+        }
+
+        [Theory]
+        [InlineData("{ CP1: var c } when c is { P1: 1 }")]
```

Comment:
This is missing because we strictly support recursive patterns (see cases above). Can be added later though.